lib/remoteproc/remoteproc_virtio.c

```diff
@@ -182,6 +225,7 @@ rproc_virtio_create_vdev(unsigned int role, unsigned int notifyid,
 			 void *rsc, struct metal_io_region *rsc_io,
 			 void *priv,
 			 rpvdev_notify_func notify,
+			 rpvdev_io_from_da_func io_from_da,
```

Comment:
The reason to introduce this function is because it needs to know how to access the the vring memory. Slave side also needs this mapping in order to access the memory. Some platform is flat, but some is not. E.g. Linux userspace. And the vring io_region can be different to the vdev config space.

Previously, when the remoteproc_create_virtio() creates the virtio device, it gets the  io region from remoteproc based on  the da address setting in the resource table.

This function is called from remoteproc function remoteproc_create_virtio() which is called directly by user.

The rproc_virtio_get_vring_info() is called from virtqueue creation, it doesn't know about the io_region as it is called from the virtio front device driver, only the remoteproc or the virtio driver knows about the mapping, but remoteproc doesn't know which mapping is for vring ahead of time.

About the code size, I also remove the old remoteproc virtio vring init function, and thus, I cannot see much code size increment from adding the API.

Reply:
This is the same point, it 's right if you decide to no more provide a virtio back end for application using remoteproc. For slave processor with limited memory,  remoteproc usage should not be mandatory. Consequence is that the io_from_da callback need to be implemented by the application to parse the memory regions. This add code but above all complexity...
To conclude, if you decide to push these patches, this means that probably need a new virtio backend based on resource table ( compatibility with linux) but not correlate to remote proc.